discovery/db_discovery.go

```diff
@@ -24,33 +24,37 @@ var getTicker = func() *time.Ticker {
 	return time.NewTicker(cacheRefreshInterval)
 }
 
+type ticketParamsValidator interface {
+	ValidateTicketParams(ticketParams *pm.TicketParams) error
+}
+
+type roundsManager interface {
+	LastInitializedRound() *big.Int
+}
+
 type DBOrchestratorPoolCache struct {
-	node *core.LivepeerNode
+	store                 common.OrchestratorStore
+	lpEth                 eth.LivepeerEthClient
+	ticketParamsValidator ticketParamsValidator
+	rm                    roundsManager
 }
 
-func NewDBOrchestratorPoolCache(node *core.LivepeerNode) *DBOrchestratorPoolCache {
-	if node.Eth == nil {
-		glog.Error("Could not refresh DB list of orchestrators: LivepeerNode nil")
+func NewDBOrchestratorPoolCache(store common.OrchestratorStore, lpEth eth.LivepeerEthClient, rm roundsManager, ticketParamsValidator ticketParamsValidator) *DBOrchestratorPoolCache {
```

Comment:
Is this really an improvement? What problem is being solved with these changes, other than the desire not to store the LivepeerNode? (Why?)

* This constructor goes from 1 parameter to 4 (and most of these params can still be derived from the original param), and the internal state goes from 1 value to 4

* We add another parameter to the public OrchestratorPool interface which means changing *all* call sites *with the same parameter that was originally removed*

* We further expand the exposed surface area of DBOrchestratorPoolCache with `CacheTranscoderPool` which drags more of the internal logic outside the DB code (and it still requires the original parameter that we removed...)

Apologies for being blunt, but this doesn't smell right to me. If there are deeper structural issues that are a concern, we can discuss, but this seems to be making things *worse* by fracturing the API all over the place.

Reply:
I see what you'r' saying but we'd still end up with 3 params in the constructor and 3 state fields even when using `LivepeerNode` so one additional param/field doesn't seem that bad to me?

We only use two of the fields on `LivepeerNode` so to me it seems that the only reason for passing in the entire object is to be able to create `Broadcaster` instances, but why create new instances when the caller (`server`) already creates these instances ?  And even if one would create  `Broadcaster` instances, it wouldn't require the entire `LivepeerNode` as a dependency because it would only use `LivepeerEthClient` methods for the time being. 

So why is `LivepeerNode` needed at all here? I see `LivepeerNode` as an aggregate of different services to operate the node, to me it feels really weird to pass in an aggregate to construct one of its the sub-services.

Reply:
Again: what's the actual problem this solves? What do we gain in exchange for fracturing the API like this?

> we'd still end up with 3 params in the constructor and 3 state fields even when using LivepeerNode

How's that? You have: the following fields in LivepeerNode: `Database`, `Eth` and `Sender` . Use those fields. Even the RoundsWatcher could be avoided if the results were written to DB.

> the only reason for passing in the entire object is to be able to create Broadcaster instances

If that's really the root of the concern, then this is much simpler:

```diff
diff --git a/core/broadcaster.go b/core/broadcaster.go
index 401f618e..33fd87f8 100644
--- a/core/broadcaster.go
+++ b/core/broadcaster.go
@@ -7,24 +7,15 @@ import (
 
 // Broadcaster RPC interface implementation
 
-type broadcaster struct {
-	node *LivepeerNode
-}
-
-func (bcast *broadcaster) Sign(msg []byte) ([]byte, error) {
-	if bcast.node == nil || bcast.node.Eth == nil {
+func (node *LivepeerNode) Sign(msg []byte) ([]byte, error) {
+	if node == nil || node.Eth == nil {
 		return []byte{}, nil
 	}
-	return bcast.node.Eth.Sign(crypto.Keccak256(msg))
+	return node.Eth.Sign(crypto.Keccak256(msg))
 }
-func (bcast *broadcaster) Address() ethcommon.Address {
-	if bcast.node == nil || bcast.node.Eth == nil {
+func (node *LivepeerNode) Address() ethcommon.Address {
+	if node == nil || node.Eth == nil {
 		return ethcommon.Address{}
 	}
-	return bcast.node.Eth.Account().Address
-}
-func NewBroadcaster(node *LivepeerNode) *broadcaster {
-	return &broadcaster{
-		node: node,
-	}
+	return node.Eth.Account().Address
 }
diff --git a/discovery/db_discovery.go b/discovery/db_discovery.go
index c009662c..8bba828f 100644
--- a/discovery/db_discovery.go
+++ b/discovery/db_discovery.go
@@ -135,7 +135,7 @@ func cacheDBOrchs(node *core.LivepeerNode, orchs []*lpTypes.Transcoder) ([]*comm
 			errc <- err
 			return
 		}
-		info, err := serverGetOrchInfo(ctx, core.NewBroadcaster(node), uri)
+		info, err := serverGetOrchInfo(ctx, node, uri)
 		if err != nil {
 			errc <- err
 			return
diff --git a/discovery/discovery.go b/discovery/discovery.go
index 6df9dbfc..6046ba21 100644
--- a/discovery/discovery.go
+++ b/discovery/discovery.go
@@ -41,8 +41,7 @@ func NewOrchestratorPool(node *core.LivepeerNode, uris []*url.URL) *orchestrator
 		randomizedUris = append(randomizedUris, uri)
 	}
 
-	bcast := core.NewBroadcaster(node)
-	return &orchestratorPool{bcast: bcast, uris: randomizedUris}
+	return &orchestratorPool{bcast: node, uris: randomizedUris}
 }
 
 func NewOrchestratorPoolWithPred(node *core.LivepeerNode, addresses []*url.URL, pred func(*net.OrchestratorInfo) bool) *orchestratorPool {
diff --git a/discovery/discovery_test.go b/discovery/discovery_test.go
index 998c22a0..c2e2b7e4 100644
--- a/discovery/discovery_test.go
+++ b/discovery/discovery_test.go
@@ -47,9 +47,8 @@ func stringsToURIs(addresses []string) []*url.URL {
 func StubOrchestratorPool(addresses []string) *stubOrchestratorPool {
 	uris := stringsToURIs(addresses)
 	node, _ := core.NewLivepeerNode(nil, "", nil)
-	bcast := core.NewBroadcaster(node)
 
-	return &stubOrchestratorPool{bcast: bcast, uris: uris}
+	return &stubOrchestratorPool{bcast: node, uris: uris}
 }
 
 func StubOrchestrators(addresses []string) []*lpTypes.Transcoder {
diff --git a/server/broadcast.go b/server/broadcast.go
index 6a09977c..739cba91 100644
--- a/server/broadcast.go
+++ b/server/broadcast.go
@@ -182,8 +182,6 @@ func selectOrchestrator(n *core.LivepeerNode, params *streamParameters, cpl core
 		return nil, errDiscovery
 	}
 
-	rpcBcast := core.NewBroadcaster(n)
-
 	tinfos, err := n.OrchestratorPool.GetOrchestrators(count)
 	if len(tinfos) <= 0 {
 		glog.Info("No orchestrators found; not transcoding. Error: ", err)
@@ -222,7 +220,7 @@ func selectOrchestrator(n *core.LivepeerNode, params *streamParameters, cpl core
 		}
 
 		session := &BroadcastSession{
-			Broadcaster:      rpcBcast,
+			Broadcaster:      n,
 			ManifestID:       params.mid,
 			Profiles:         params.profiles,
 			OrchestratorInfo: tinfo,

```

Reply:
In your example we wouldn't have to pass `node` as a param to `serverGetOrchInfo` because we already have an instance of it on `DBOrchestratorPoolCache`, right? 

> How's that?
I forgot about pm.Sender (now ticketParamsValidator), you're right. 

> If that's really the root of the concern

It's not the root of the concern, it's a consequence of the root of the concern. How I see it `Broadcaster` and `Orchestrator` are specific instances of a general `LivepeerNode` instance that aggregates different functionalities from several packages , both with their specific exposed methods through respective interfaces. 

`DBOrchestratorPool` is not a specific instance of `LivepeerNode`, it is a service running on `LivepeerNode`. No other of these 'services'  bound to it use `LivepeerNode` in their constructor or have it as state. Only `DBOrchestratorPoolCache` does, so to me personally that isn't logical. Or is that a wrong perspective? 

There's also an idiom in go that says "pass interfaces, return structs" which I think roughly translates to the "interface segregation principle" described in [solid go design - dave cheney](https://dave.cheney.net/2016/08/20/solid-go-design)

Lastly I don't think a high-level package such as `core` should be imported in a lower level package such as `discovery`  , not having `core.LivepeerNode` as a dependency solves that.

Reply:
>  their specific exposed methods through respective interfaces.

Would you feel better if we added `Database()` , `Eth()` and `Sender()` methods to LivepeerNode (or Broadcaster) and just defined some interface for `NewDBOrchestratorPoolCache` taking all those? (Not that I actually think it's a good idea; it smells too much like Java. And being too prescriptive about design patterns is exactly why people make fun of Java.)

> No other of these 'services' bound to it use LivepeerNode in their constructor or have it as state. Only DBOrchestratorPoolCache does, so to me personally that isn't logical. Or is that a wrong perspective?

You say it's the only "service" that uses LivepeerNode. OK, so what actual issues has this led to? Why is DBOrchestratorPoolCache using LivepeerNode a problem? (Again, this is the original question.)

I'm also not sure which other services you're specifically referring to, but some of the other stuff is heavily parameterized by interfaces - almost excessively so IMO. Perhaps those could be tidied up by a good helping of LivepeerNode.

> roughly translates to the "interface segregation principle" described in solid go design - dave cheney

It is fine and good to recognize certain classes of design patterns when they're appropriate, but dogmatic adherence leads to... this.

I dislike using design patterns as a cudgel during discussion because that tends to ignore the practical realities of the system in question, so I don't bring up design patterns often, but here is one that's being broken all over the place: favor [deep abstractions over shallow ones](https://web.stanford.edu/~ouster/cgi-bin/cs190-winter18/lecture.php?topic=modularDesign) [1]. This is why I often refer to the 'surface area' of an API - and right now, things are blowing up.

Not only is the OrchestratorPool abstraction being expanded, DBOrchestratorPool has its polling/caching exposed to external callers, and the complexity of initializing everything is hoisted out to `livepeer.go`.

[1] Ousterhout's "A Philosophy of Software Design" has much more on this, and the ideas in there have found their way into a lot of the golang community's recommendations.

> Lastly I don't think a high-level package such as core should be imported in a lower level package such as discovery

Dunno, `core` gives me the impression that it's, well, core - in the middle of everything?

![image](https://user-images.githubusercontent.com/292510/69467699-0e767880-0d3e-11ea-9fdb-b71881042cfc.png)

Reply:
> You say it's the only "service" that uses LivepeerNode. OK, so what actual issues has this led to? 

> This is why I often refer to the 'surface area' of an API - and right now, things are blowing up.

Couldn't one make a similar argument for passing LivepeerNode as a concrete type as an argument? As it is likely to expose a lot of functionality that the package in question does not need. 

> Not only is the OrchestratorPool abstraction being expanded, DBOrchestratorPool has its polling/caching exposed to external callers, and the complexity of initializing everything is hoisted out to livepeer.go.

The interface remains the same (except for the slight signature change), these methods aren't part of the interface so any package that uses `OrchestratorPool` has no knowledge of those methods. Whether we cache or not is depending on whether the last seen block is within the current round or not. But I'd be happy to learn if there is a better pattern here. 

> Dunno, core gives me the impression that it's, well, core - in the middle of everything? 

Yes I agree (altough DB is actually in the middle in your pic), but I see it partially different. 
(diagram incomplete and not completely accurate) 
![image](https://user-images.githubusercontent.com/22256858/69504331-21e23900-0f22-11ea-8831-a4f4d720dc62.png)


Anyway design patterns aside I'm wondering where we can find some common ground in this discussion.  Would you be more okay with the following? 

```
type DBOrchestratorPoolCache struct {
	store                 common.OrchestratorStore
	lpEth                 eth.LivepeerEthClient
	ticketParamsValidator ticketParamsValidator
	rm                    roundsManager
	bcast common.Broadcaster
}

func NewDBOrchestratorPoolCache(n *core.LivepeerNode, rm roundsManager) *DBOrchestratorPoolCache {
	bcast := core.NewBroadcaster(n)
	return &DBOrchestratorPoolCache{
		store: n.Database,
		lpEth: n.Eth,
		ticketParamsValidator: n.Sender,
		rm: rm,
		bcast: bcast,
	}
}

type OrchestratorPool interface {
	GetURLs() []*url.URL
	GetOrchestrators(int) ([]*net.OrchestratorInfo, error)
	Size() int
}
```

Reply:
> Couldn't one make a similar argument for passing LivepeerNode as a concrete type as an argument? As it is likely to expose a lot of functionality that the package in question does not need.

Having the functionality generally available but only used by other packages is not really a problem. What's problematic is *how* the exposed functionality is expected to be used. As-is right now, too much is being expected of API callers in terms of set-up logic, when most of that logic could be pulled into the DB cache itself. APIs should be misuse-resistant [1] to minimize the risk of "user error" but the current changes substantially increase that risk.

[1] Section 2.3 of the [API Design Manual](https://people.mpi-inf.mpg.de/~jblanche/api-design.pdf), "Hard to misuse"

> The interface remains the same (except for the slight signature change)

The signature change to OrchestratorPool is *significant*, not slight. You're now making it a requirement that the node / broadcaster is available at all call sites of `GetOrchestrator`. Currently that's not really a problem but it may not always be. That is an avoidable issue and the change is unnecessary.

Then there's the change to the usage of `DBOrchestratorPoolCache`, where the caller (`livepeer.go`) needs to do a lot of manual set-up, *and* the particulars of the caching/polling setup are such that it's actually sensitive to initialization order [1]. Both issues are also avoidable, because none of that actually needs to take place outside the `NewDBOrchestratorPoolCache`. And the new functions require essentially the same parameters - the `Broadcaster` - which is yet another indication things should be reworked.

[1] https://github.com/livepeer/go-livepeer/pull/1182#discussion_r349409217
[2] https://people.mpi-inf.mpg.de/~jblanche/api-design.pdf

> Would you be more okay with the following?

My suggestion is to keep overall API changes as small as possible. From what I can tell, the only change that may be required for the functional purpose of this PR is the inclusion of the rounds manager to the `NewDBOrchestratorPoolCache`. So that means:

* Revert the change to OrchestratorPool.GetOrchestrators
* Add the `roundsManager` to `NewDBOrchestratorPoolCache`
* Initialize the cache / poller *within* the `NewDBOrchestratorPoolCache` . This doesn't need to be done within `livepeer.go`.

It's a bit of a sport to keep diffs as minimal as possible while achieving the goals of the task. How small can you go?

Reply:
IIUC, the following points are being debated:

1. The modification of the public API of the OrchestratorPool interface (change to the signature of GetOrchestrators())
2. The larger API surface area of DBOrchestratorPoolCache because the caller is now responsible for transcoder pool caching and polling via the CacheTranscoderPool() and PollOrchestratorInfo() methods respectively
3. The injection of 3 interface dependencies into DBOrchestratorPoolCache instead of injecting a single concrete LivepeerNode dependency

My take is:

1. +1 for not changing the OrchestratorPool public API since updating the public API by passing an additional arg in the method has a clear cost due to the requirements for call site changes. Furthermore, the caller now needs to be aware of this dependency which might be a worse state of affairs than DBOrchestratorPoolCache being aware of the dependency - we go from a single concrete type being aware of a dependency to every single type that implements the interface being aware of a depdency

2. +1 that the caller shouldn't be responsible for setting up caching and polling logic. We could either move this into the constructor or into a dedicated method i.e. `Start()` that accepts a cancellable context (that can be used to exit the polling goroutine gracefully)

3. I think Josh makes a good point about focusing on how functionality is used instead of just exposed functionality. I also understand the desire to abstract away the implementation details of dependencies by *only* using interfaces. I think avoiding introducing additional package dependencies when we can is a good thing (in this case the `core` package). However, in this case, I think reaching for these abstractions might be premature and we should try to avoid reaching for abstractions until we actually need them rather than trying to predict their need. There is a saying: [duplication is cheaper than the wrong abstraction](https://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction) - in this case I think accepting the current state of dependencies is cheaper than reaching for what *could* be the wrong abstraction in the absence of more information. For that reason, I'm also +1 on this point

Reply:
Let's get this PR merged first a foremost and we can always discuss this later if we want. I'd just want to close off by saying that I can agree with 1 & 2 but not in full with point 3. I listed my objections below just for completeness. But let's get 6061820 reviewed and I hope everyone can find themselves happy with the changes in that commit. 

- Both types are now dependant of eachother, being one step away from cyclical imports
- `LivepeerNode` is not needed here at all, simply saying "it's there, just don't use it"  doesn't feel right either. 
- Importing an additional package just because it adds a few constructor parameters (what's wrong with composition?) is in my opinion the wrong trade-off to make
- I agree that we shouldn't prematurely reach for abstractions but on the other hand whether an abstraction is wrong can only be known in hindsight, so the statement made by the author there is actually a perfect example of a petitio principii (one can equally say "the right abstraction is cheaper than duplication").

Reply:
> so the statement made by the author there is actually a perfect example of a petitio principii (one can equally say "the right abstraction is cheaper than duplication").

I think the author's point is being misinterpreted a bit here...Of course we only know if an abstraction is wrong in hindsight. The point is that *if* the abstraction turns out to be wrong, then dealing with the consequences of that abstraction is almost always more of a headache than dealing with duplication. And "duplication" here can probably be replaced by a variety of not-so-ideal pieces of code including code that exposes more functionality than needed.